### ISSUE TITLE
[RFC] Feature: support Active Learning training loop

### ISSUE THREAD

**Issue by pietrolesci** [PyTorchLightning/lightning-flash#761]
# Feature 🚀: Active Learning training loop

Active learning is a special case of machine learning in which a learning algorithm is able to
interactively query the user (or some other information source) to obtain the desired outputs
at new data points.

In practice, it is often implemented as "pool-based" active learning: data is readily available
but labelling is costly. The active learning process proceeds as follows. Assuming you have a 
unlabelled dataset (a "pool") and you labelled (or "acquired") (maybe randomly) some inputs from the pool

1. Train the model for one epoch on those labelled data (within each epoch the data can be seen multiple
times - this is important especially when the initial labelled dataset is small; of course how
many times the dataset is repeated within each epoch can vary through the active training)

1. Evaluate the model on the pool -> (computationally intensive, maybe use the new `torch.inference_mode()`
context manager? Maybe use something like [toma](https://github.com/BlackHC/toma) to max out memory space?)

1. Feed the model predictions/statistics (depends on the specific active learning strategy) to
an "acquisition"/__scoring function__ which scores the inputs in the pool

1. Once the scores are available, apply a __query strategy__ (e.g., top _k_, max margin, least confidence)
that selects the actual inputs to label

1. Label inputs from the pool and add it to the training set -> (might be problematic in distributed setting)

1. (Optional) Re-initialize the model from scratch (weights, scheduler, optimizer)

In the current state of Flash, this can be achieved via a callback. However, this poses challenges
in the distributed setting, makes it difficult to reset states, and makes it more difficult to work
with existing `training_strategies`.

## Proposed API
The alternative is to implement active learning in Flash via a new trainer, the `ActiveLearningTrainer`.
This new trainer would provide access to the `ActiveLearningLoop` ... [here is for you @tchaton :)]

This is justified by the fact that

- Active learning is effectively an outer loop on top of the standard
training loop

- User don't have to learn a new API as it will be the same as the current trainer

- Researchers can easily hack into the active learning loop

## Desiderata

From a researcher point of view, it is important to be able to hack the following things

- How many times the training data are repeated within each epoch and be able to change that, much like a scheduler

- Change how the model is evaluated on the pool (e.g., certain AL strategies need many forward passes on the same
data to estimate uncertainty)

- Change the acquisition (aka scoring) functions

- Change the query strategy
